src/battle-log.ts

```diff
@@ -222,6 +249,30 @@ class BattleLog {
 			break;
 		}
 	}
+	displayJoinLeaves(preList: string[], action: string) {
+		let message = '';
+		const list: string[] = [];
+		for (const user of preList) {
+			if (!list.includes(user)) list.push(user);
+		}
+		for (let j = 0; j < list.length; j++) {
+			if (j >= 5) {
+				message += `, and ${list.length - 5} others`;
+				break;
+			}
+			if (j > 0) {
+				if (j === 1 && list.length === 2) {
+					message += ' and ';
+				} else if (j === list.length - 1) {
+					message += ', and ';
+				} else {
+					message += ', ';
+				}
+			}
+			message += BattleLog.escapeHTML(list[j]);
+		}
+		return `${message} ${action}`;
+	}
```

Comment:
I realize I'm nitpicking, but I think this would be better at a slightly different level of abstraction:

```suggestion
	textList(list: string[]) {
		let message = '';
		const listNoDuplicates: string[] = [];
		for (const user of list) {
			if (!listNoDuplicates.includes(user)) listNoDuplicates.push(user);
		}
		list = listNoDuplicates;
		for (let i = 0; i < list.length; i++) {
			if (i >= 5) {
				message += `, and ${list.length - 5} others`;
				break;
			}
			if (i > 0) {
				if (i === 1 && list.length === 2) {
					message += ' and ';
				} else if (i === list.length - 1) {
					message += ', and ';
				} else {
					message += ', ';
				}
			}
			message += list[i];
		}
		return message;
	}
```